src/dune_rules/dune_file.ml

```diff
@@ -834,10 +835,18 @@ module Library = struct
     let virtual_library = is_virtual conf in
     let foreign_archives = foreign_lib_files conf ~dir ~ext_lib in
     let native_archives =
-      if modes.native then
-        [ archive ext_lib ]
+      let archive = archive ext_lib in
+      if virtual_library || not modes.native then
+        Lib_info.Files []
+      else if
+        Option.is_some conf.implements
+        || Lib_config.linker_can_create_empty_archives lib_config
+           && Ocaml_version.ocamlopt_always_calls_library_linker
+                lib_config.ocaml_version
```

Comment:
It seems that this logic should be reproduced in `Lib_info.eval_native_archives`

Reply:
Hmm, but why? If we create a `Lib_info.t` from a dune-package, all of this logic is avoided as the native archives are evaluated. I might be misunderstanding this comment.

Reply:
Agreed, but it seems to me that we call `eval_native_archives` for `Lib_info.t` values not coming from a `dune-package` file. For instance in `install_rules.ml`

Reply:
Basically my mental model is as follow:

in some cases, we immediately know the set of modules. In some cases, we have to do a bit of computation before we know it. The latter case correspond to `Needs_module_info`.

The fact that a library has a native archive depends on two things:
- whether the library has a least one implementation
- the version of OCaml

so it seems to me that whenever we compute the list of native archives, we should be looking at this two information. However, looking at the current code, we only look at the version of OCaml when the list of modules is know statically. That doesn't seem right. Do you agree?

Reply:
> The fact that a library has a native archive depends on two things:

Not quite. We only need module information only *if* the version of OCaml cannot tell us this information. So essentially:

```
ocaml_version > 4.12   => Needs_module_info archive
ocaml_version <= 4.12 => Files archive
```

> so it seems to me that whenever we compute the list of native archives, we should be looking at this two information. However, looking at the current code, we only look at the version of OCaml when the list of modules is known statically.

We look at the version of OCaml version for all locally defined libraries whenever we cannot guarantee the existence of the native archive due to some other condition. 

> Agreed, but it seems to me that we call eval_native_archives for Lib_info.t values not coming from a dune-package file. For instance in install_rules.ml

Right. For such libraries, the OCaml version information has already been factored in when constructing `Lib_info.t` from `Dune_file.Library.t`. We already evaluated

```ocaml
        || Lib_config.linker_can_create_empty_archives lib_config
           && Ocaml_version.ocamlopt_always_calls_library_linker
                lib_config.ocaml_version
```

if necessary.

Reply:
Ack, i understand now